Interface

Crablicious opened the pull request:
Pull request, with the hopes that code lines can be commented for code review.

dowenberghmark committed:
added double buffer

dowenberghmark committed:
added Makefile

dowenberghmark committed:
created class double buffer

dowenberghmark committed:
function skelett

Crablicious committed:
defs for buffer

dowenberghmark committed:
created a cpu dummy

dowenberghmark committed:
wrong function name

Crablicious committed:
Dummy fpga running in its own thread. Tested with three manual iterations from a manually created double buffer

Crablicious committed:
Merge branch 'interface' of github.com:dowenberghmark/PineappleExpress into interface

saraback committed:
Double buffer stuff. Needs to be checked

saraback committed:
Merge branch 'interface' of github.com:dowenberghmark/PineappleExpress into interface

saraback committed:
Indentation issues

Crablicious commented on interface/cpu_dummy.cpp:
```diff
@@ -0,0 +1,51 @@
+                                    "5555555555555555555555555555555555555555555555555555555555555555"};
+
+
+  //adding two strings to the buffer and start processing
```
This section should probably be refactored into a function, such that its easy to change as we change the API. Also gives better overview of the behavior.

saraback committed:
Implementation of double buffer. Needs a review

Crablicious reviewed (commented):
Should work, minor refactoring for better readability only.

saraback committed:
Code standard changes for pointer variables

dowenberghmark commented on interface/fpga.hpp:
```diff
@@ -0,0 +1,73 @@
+
+		buffers[0] = (struct buffer *) buffer0;
+		buffers[1] = (struct buffer *) buffer1;
+		buffers[0]->chunks = (struct chunk *)((char *) buffer0 + sizeof(struct buffer_header));
```
Just an implementation question. Should the FPGA have constant offsets so it can fetch different data accordingly?

dowenberghmark commented on interface/fpga.hpp:
```diff
@@ -0,0 +1,73 @@
+		*ready_flag = 1;
+	}
+}
+
```
Should be fine otherwise

Crablicious committed:
main -> fpga_main, 2 space indent and define for buffer header size

saraback committed:
More code style changes

saraback committed:
Merge branch 'interface' of github.com:dowenberghmark/PineappleExpress into interface

saraback committed:
Added destructor to Double_buffer.cpp

Crablicious committed:
fpga implementation extracted into fpga.cpp

dowenberghmark committed:
Refactored some parts into a function

dowenberghmark committed:
Merge branch 'interface' of github.com:dowenberghmark/PineappleExpress into interface

saraback committed:
Added logic to get_result() in Double_buffer.cpp

saraback committed:
Merge is needed

saraback commented on interface/cpu_dummy.cpp:
```diff
@@ -0,0 +1,38 @@
+  chunk_placement_ptr = double_buffer->get_chunk();
+  strcpy(chunk_placement_ptr, current.c_str());
+}
+void deploy_work(int total_chunks, int chunks_per_buffer, std::string* data, Double_buffer* our_double_buffer){
```
Seems good. A few style changes to pointer variables are needed for compliance to code style. 

`int *foo` instead of `int* foo`

dowenberghmark committed:
changed to the right code std

Crablicious commented on interface/double_buffer.hpp:
```diff
@@ -0,0 +1,33 @@
+class Double_buffer
+{
+ private:
+  char *first_buffer; //start of first buffer
```
Would be good with consistent naming, such that they follow the same structure. first_buffer / buf_head_first and second_buffer / buf_head_second collides.

dowenberghmark committed:
wrong type and destructor doesn't have default

Crablicious commented on interface/double_buffer.hpp:
```diff
@@ -0,0 +1,33 @@
+ private:
+  char *first_buffer; //start of first buffer
+  char *second_buffer; //start of second buffer
+  char *place_to_write; //place to write to in active buffer
```
Something more specified would be good, perhaps chunk_to_write or current_chunk or chunk_ptr.

Crablicious commented on interface/double_buffer.cpp:
```diff
@@ -0,0 +1,70 @@
+#include "defs.hpp"
+
+
+Double_buffer::Double_buffer(uint32_t chunks){
```
I think the pointer arithmetic in this section is not working as intended.
For example:
`buf_head_first = (buffer_header *) (glob_head + sizeof(global_header));`
 
`glob_head` is of the type `struct global_header *` so `glob_head + 1` would point to the next `struct global_header` object. `glob_head + sizeof(global_header)` would jump forward 8*8 bytes instead of the intended 8 bytes. Either change to glob_head + 1 or cast glob_head to a `char *` pointer to get the intended behavior.

I'll mark the perpetrators.

Crablicious commented on interface/double_buffer.cpp:
```diff
@@ -0,0 +1,70 @@
+  glob_head -> start_processing_flag = 0;
+  glob_head -> active_buffer_flag = 0;
+
+  buf_head_first = (buffer_header *) (glob_head + sizeof(global_header));
```
!

Crablicious commented on interface/double_buffer.cpp:
```diff
@@ -0,0 +1,70 @@
+  buf_head_first -> num_chunks = 0;
+  buf_head_first -> ready_flag = 0;
+
+  buf_head_second = (buffer_header *) (buf_head_first + sizeof(buffer_header) + sizeof(chunk) * chunks);
```
!

Crablicious commented on interface/double_buffer.cpp:
```diff
@@ -0,0 +1,70 @@
+  buf_head_second -> num_chunks = 0;
+  buf_head_second -> ready_flag = 0;
+
+  first_buffer = (char *) (buf_head_first + sizeof(buffer_header));
```
!

Crablicious commented on interface/double_buffer.cpp:
```diff
@@ -0,0 +1,70 @@
+  buf_head_second -> ready_flag = 0;
+
+  first_buffer = (char *) (buf_head_first + sizeof(buffer_header));
+  second_buffer = (char *) (buf_head_second + sizeof(buffer_header));
```
!

Crablicious reviewed (commented):
Some pointer arithmetics that need to be looked at, otherwise good. Perhaps swap all sizeof to constants defined in defs.h? Got BUFFER_HEADER_SIZE there atm.

saraback committed:
Added size defs for buffer head and chunk size

saraback committed:
Pointer arithmetic fixed, name changes for consistency

saraback committed:
Pointer arithmetic hopefully fixed for real this time

Crablicious commented on interface/double_buffer.cpp:
```diff
@@ -0,0 +1,70 @@
+  second_buf_head -> num_chunks = 0;
+  second_buf_head -> ready_flag = 0;
+
+  first_buf = (char *) (first_buf_head + BUFFER_HEADER_SIZE);
```
Parenthesis needs to be removed for this to work. Same on the line under (l. 24).

saraback committed:
More arithmetic stuff

dowenberghmark committed:
updated cmake for integration

dowenberghmark committed:
Merge branch 'interface' of github.com:dowenberghmark/PineappleExpress into interface

dowenberghmark committed:
got a bug in accesing th first buffer

Crablicious committed:
struct buffer no longer casted to real buffer. some indents

dowenberghmark committed:
BUG--Can only change the double buffer twice

Crablicious committed:
buffer is now flipping correctly

saraback committed:
Code refactoring

Crablicious committed:
refactoring. BUFFER_COUNT introduced.

dowenberghmark committed:
refractored the cpu_dummy into a class

dowenberghmark committed:
some merges

saraback committed:
Added picture definition of DoubleBuffer

saraback committed:
Merge branch 'interface' of github.com:dowenberghmark/PineappleExpress into interface

saraback committed:
Added directory for diagrams, with subdirectories for generated pdf
and for the xml files to open in draw.io

Crablicious committed:
posix calls are used to communicate with fpga now. saved in dram.hex,… (#2)

Posix calls are used to communicate with fpga now. saved in dram.hex, which can be inspected in an hex editor of your liking. Documentation about buffer usage is written in double_buffer.cpp. Also some refactoring of names.

saraback committed:
Updated diagram with flag switch

Crablicious committed:
strings should be 64 bytes, not 63

Crablicious committed:
set c++ 11 in cmake

Snapple49 commented on interface/CMakeLists.txt:
```diff
@@ -0,0 +1,7 @@
+cmake_minimum_required(VERSION 3.1)
+project( interface )
```
need to add "set(CMAKE_CXX_STANDARD 11)"

Snapple49 commented on interface/cpu_dummy.cpp:
```diff
@@ -0,0 +1,95 @@
+  }
+}
+
+void InterfaceTest::flipping_buffer(int amount_buffer_flips, int number_of_strings){
```
the name flipping_buffer is not very clear

Snapple49 replied:
name flipping_buffer is not very clear

Snapple49 commented on interface/cpu_dummy.cpp:
```diff
@@ -0,0 +1,95 @@
+
+void InterfaceTest::flipping_buffer(int amount_buffer_flips, int number_of_strings){
+  char *result;
+  bool flag;
```
could have more descriptive name for flag

saraback committed:
Added buffer intraction sequence diagram

saraback committed:
Merge branch 'interface' of github.com:dowenberghmark/PineappleExpress into interface

Snapple49 commented on interface/cpu_dummy.cpp:
```diff
@@ -0,0 +1,95 @@
+      }else{
+        strncpy(chunk_placement_ptr, string_array[counter_chunks % number_of_strings].c_str(), 64);
+      }
+      counter_chunks++;
```
should counter_chunks really increment when placement pointer is null?

dowenberghmark replied:
Yeah I did something right on the other branch i deleted

Crablicious committed:
to remove warnings

Crablicious committed:
Merge branch 'interface' of github.com:dowenberghmark/PineappleExpress into interface

Snapple49 commented on interface/double_buffer.cpp:
```diff
@@ -0,0 +1,133 @@
+  }
+
+  glob_head->start_proc = 0;
+  glob_head->active_buf = 1;
```
maybe add comment about indexing, does 1 mean buffer1 or buffer0, and also explain values of start_proc, valid values? Or add defines?

Snapple49 commented on interface/fpga.cpp:
```diff
@@ -0,0 +1,88 @@
+  int ret;
+  ret = pread_all(dram_fd, &glob_head, GLOBAL_HEADER_SIZE, dram_glob_head);
+  if (ret < 0) {
+    throw std::runtime_error("Can't write to dram from fpga.");
```
Copy paste error message, wrong? Or is write_glob error msg wrong?

Snapple49 commented on interface/fpga.cpp:
```diff
@@ -0,0 +1,88 @@
+  // only read the chunks that have been written to.
+  ret = pread_all(dram_fd, &buf.data, CHUNK_SIZE*buf.head.num_chunks, dram_bufs[glob_head.active_buf]);
+  if (ret < -1) {
+    throw std::runtime_error("Can't write to dram from fpga.");
```
More error msg mistakes

Snapple49 commented on interface/fpga_main.cpp:
```diff
@@ -0,0 +1,53 @@
+  t.join();
+
+  return 0;
+}
```
THIS FILE IS USELESS SHIT

Snapple49 replied:
SHAME ON YOU

Snapple49 reviewed (changes requested):
Looks good, some names not very descriptive, some values could maybe use definitions/comments, copy-paste errors

saraback committed:
Fixes added after code review

Crablicious committed:
removed legacy testing main

Crablicious committed:
Merge branch 'interface' of github.com:dowenberghmark/PineappleExpress into interface

Crablicious committed:
Aws interface (#6)

Moves from cmake to Makefile, since AWS provides a Makefile that works and that we can modify to our needs easily.

The system can now be compiled either for AWS (without fpga dummy) or for local (with fpga dummy) using 'make mode=aws' or 'make mode=local'.
